libz-rs-sys/src/gz.rs

```diff
@@ -607,6 +707,228 @@ pub unsafe extern "C-unwind" fn gzeof(file: gzFile) -> c_int {
+    if unsafe { inflateInit2(&mut state.stream as *mut z_stream, MAX_WBITS + 16) } != Z_OK {
+        // Safety: The caller confirmed the validity of `state`, and `free_buffers` checks
+        // for null input and output pointers internally.
+        unsafe { free_buffers(state) };
+        // Safety: The caller confirmed the validity of `state`.
+        unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+        return Err(());
+    }
+
+    // Get at least the magic bytes in the input buffer.
+    if state.stream.avail_in < 2 {
+        // Safety: The caller confirmed the validity of `state`.
+        let Ok(_) = (unsafe { gz_avail(state) }) else {
+            return Err(());
+        };
+        if state.stream.avail_in == 0 {
+            return Ok(());
+        }
+    }
+
+    // Look for gzip magic bytes.
+    // Safety: `gz_avail` ensures that `next_in` points to at least `avail_in` readable bytes.
+    if state.stream.avail_in > 1
+        && unsafe { *state.stream.next_in } == 31
+        && unsafe { *state.stream.next_in.add(1) } == 139
+    {
+        // Safety: We initialized `state.stream` with `inflateInit2` above.
+        unsafe { inflateReset(&mut state.stream as *mut z_stream) };
+        state.how = How::Gzip;
+        state.direct = false;
+        return Ok(());
+    }
+
+    // No gzip header. If we were decoding gzip before, the remaining bytes
+    // are trailing garbage that can be ignored.
+    // FIXME: Add test coverage for this case (which may require the remaining decompression logic).
+    if !state.direct {
+        state.stream.avail_in = 0;
+        state.eof = true;
+        state.have = 0;
+        return Ok(());
+    }
+
+    // The file is not in gzip format, so enable direct mode, and copy all
+    // buffered input to the output.
+    // Safety:
+    // * `state.output` was allocated above.
+    // * `gz_avail` ensures that `next_in` points to at least `avail_in` readable bytes.
+    unsafe {
+        ptr::copy(
+            state.stream.next_in,
+            state.output,
+            state.stream.avail_in as usize,
+        )
+    };
+    state.next = state.output;
+    state.have = state.stream.avail_in;
+    state.stream.avail_in = 0;
+    state.how = How::Copy;
+    state.direct = true;
+
+    Ok(())
+}
+
+// Load data into the input buffer and set the eof flag if the last of the data has been
+// loaded.
+//
+// # Returns
+//
+// * `Ok(n)` on success, where `n` is the number of bytes available (`state.stream.avail_in`)
+// * `Err` on error
+//
+// # Safety
+//
+// `state` must have been properly initialized, e.g. by [`gzopen_help`].
+unsafe fn gz_avail(state: &mut GzState) -> Result<usize, ()> {
+    if state.err != Z_OK && state.err != Z_BUF_ERROR {
+        return Err(());
+    }
+    if !state.eof {
+        if state.stream.avail_in != 0 {
+            // Copy any remaining input to the start.
+            unsafe {
+                ptr::copy(
```

Comment:
I suspect also here 
```suggestion
                ptr::copy_nonoverlapping(
```

Reply:
In this case, we're shifting data from the middle of a buffer to the front, so it may overlap.

Reply:
Maybe add a note then that they may overlap